syntax = "proto3";

package lol_sounds;

message Event {
  enum Side {
    SIDE_UNSPECIFIED = 0;
    BLUE = 1;
    RED = 2;
  }
  Side side = 1;

  enum Scenario {
    UNSPECIFIED = 0;

    // Frequent scenario will take priority numbers 1 - 15
    // Kills
    KILL = 1;
    FIRST_BLOOD = 2;
    SHUTDOWN = 10;
    DOUBLE = 8;
    TRIPLE = 9;
    QUADRA = 14;
    PENTA = 15;
    SPREE = 11;
    RAMPAGE = 12;
    UNSTOPPABLE = 13;
    DOMINATING = 16;
    GODLIKE = 17;
    LEGENDARY = 18;

    // Objectives
    TURRET = 3;
    INHIBITOR = 4;
    DRAGON = 6;
    BARON = 7;

    // Miscellaneous
    GAME_END = 5;
  }
  Scenario scenario = 2;

  Participant.ID participant_id = 3;
}

// A snapshot represents a time when an event(s) take place
message Snapshot {
  // Indicates the in game time in seconds when the events took place.
  double game_time = 1;
  repeated Event events = 2;
}

message Participant {
  // The esports player id.
  int64 player_id = 1;

  // Game specific details
  string champion = 2;

  enum ID {
    ID_UNSPECIFIED = 0;

    BLUE_TOP = 1;
    BLUE_JUNGLE = 2;
    BLUE_MID = 3;
    BLUE_BOT = 4;
    BLUE_SUPPORT = 5;

    RED_TOP = 6;
    RED_JUNGLE = 7;
    RED_MID = 8;
    RED_BOT = 9;
    RED_SUPPORT = 10;
  }
  ID id = 3; // The participant id

  enum Role {
    ROLE_UNSPECIFIED = 0;

    TOP = 1;
    JUNGLE = 2;
    MID = 3;
    BOTTOM = 4;
    SUPPORT = 5;
  }
  Role role = 4;

  string summoner_name = 5;
}

message Metadata {
  int64 game_id = 1;
  int64 match_id = 2;

  // Empty patch string means that stats are not available
  string patch = 3;

  message Team {
    int64 id = 1;
    string name = 2; // The full name of the team.
    string code = 3; // The team's name in short form.

    repeated Participant participants = 4;
  }
  Team blue_team = 4;
  Team red_team = 5;
}

message Game {
  Metadata metadata = 1;

  // The time interval in seconds between TimeFrames
  int32 time_interval = 2;

  message Vod {
    // The video id
    string id = 1;
    // The platform the vod is on. (YouTube, Twitch, Znipe e.t.c)
    string platform = 2;

    // These represent the video player times in seconds
    double start_time = 3;
    double end_time = 4;
  }
  repeated Vod vods = 3;

  message TimeFrame {
    repeated Snapshot snapshots = 1;
  }
  repeated TimeFrame frames = 4;
}